Fix missing background on upgrade

Pytal opened the pull request:
Fix ~~https://github.com/nextcloud/server/issues/34458~~ https://github.com/nextcloud/server/issues/34216

Background settings previously were set in the `oc_preferences` table with the appid `"dashboard"`, after https://github.com/nextcloud/server/pull/33733 the settings were queried for with the appid `"theming"`

This migration aims to fix the background loss on upgrade by updating the appid to `"theming"`

Pytal committed:
Fix missing background on upgrade

Signed-off-by: Christopher Ng <chrng8@gmail.com>

Pytal commented:
/backport to stable25

Pytal commented:
The migration looks right but doesn't seem to work, any ideas why @blizzz?

nickvergessen commented on core/Migrations/Version25000Date20221007010957.php:
```diff
@@ -0,0 +1,65 @@
+		$qb->update('preferences')
+			->set('appid', 'theming')
+			->where($qb->expr()->eq('appid', $qb->createNamedParameter('dashboard')))
+			->andWhere($qb->expr()->eq('configkey', $qb->createNamedParameter('background')))
+			->orWhere($qb->expr()->eq('configkey', $qb->createNamedParameter('backgroundVersion')));
```
```suggestion
			->andWhere($qb->expr()->orX(
				$qb->expr()->eq('configkey', $qb->createNamedParameter('background')),
				$qb->expr()->eq('configkey', $qb->createNamedParameter('backgroundVersion'))
			));
```

nickvergessen commented on core/Migrations/Version25000Date20221007010957.php:
```diff
@@ -0,0 +1,65 @@
+ */
+class Version25000Date20221007010957 extends SimpleMigrationStep {
+
+	/** @var IDBConnection */
+	protected $connection;
```
```suggestion
	protected IDBConnection $connection;
```

nickvergessen commented on core/Migrations/Version25000Date20221007010957.php:
```diff
@@ -0,0 +1,65 @@
+		$qb = $this->connection->getQueryBuilder();
+
+		$qb->update('preferences')
+			->set('appid', 'theming')
```
```suggestion
			->set('appid', $qb->createNamedParameter('theming'))
```

Exception: Database error when running migration 25000Date20221007010957 for app core An exception occurred while executing a query: SQLSTATE[42S22]: Column not found: 1054 Unknown column 'theming' in 'field list'

nickvergessen committed:
Fix SQL errors

Signed-off-by: Joas Schilling <coding@schilljs.com>

nickvergessen committed:
New code being new

Signed-off-by: Joas Schilling <coding@schilljs.com>

PVince81 commented on core/Migrations/Version25000Date20221007010957.php:
```diff
@@ -0,0 +1,65 @@
+		$qb = $this->connection->getQueryBuilder();
+
+		$qb->update('preferences')
+			->set('appid', $qb->createNamedParameter('theming'))
```
assuming no collision / key violation possible ?
should we add an extra where to guarantee that there is not already one under the "theming" name ?

Pytal replied:
Migrations should not address states introduced due to running on `master` (not released) versions right @nickvergessen?

As this will be backported to 25 and the switch from dashboard to theming is also 25

nickvergessen replied:
I guess you have to take care of it, because otherwise the update breaks e.g. on our instance when people configured an background already?

PVince81 replied:
it looks like it would need rewriting the query and/or using a loop with maybe catching exceptions
might be too much of a hassle to cover for a potential edge case that might not even happen ?

PVince81 replied:
the issue would only affect users of RC3+ anyway, so maybe we delete all the "theming" keys before migrating the "dashboard" ones, assuming that no one pre-RC3 have the "theming" key in place anyway.

for those post-RC3 it might be ok to ask them to set the background again ? (they were using a pre-release anyway!)

nickvergessen replied:
Also fine with deleting, but needs to be done before we can merge this.

Pytal committed:
Delete theming background preferences

Signed-off-by: Christopher Ng <chrng8@gmail.com>

Pytal commented:
Probably just my local instance not working but updated and ready for review

PVince81 reviewed (approved):
:+1: as discussed

PVince81 commented:
@blizzz we need this for 25.0.0RC4

come-nc reviewed (approved):
I would feel better if it checked theming/background is not set before overwriting it.

PVince81 commented:
> I would feel better if it checked theming/background is not set before overwriting it.

this is what we wanted to avoid as it would require iterating over all users and slow down the migration/update

in normal cases the migration would not rerun

blizzz commented on core/Migrations/Version25000Date20221007010957.php:
```diff
@@ -0,0 +1,71 @@
+			->where($qb->expr()->eq('appid', $qb->createNamedParameter('theming')))
+			->andWhere($orExpr);
+
+		$qb->update('preferences')
```
get a new `QueryBuilder` instant, this will have state from the previous operation.

PVince81 replied:
will this break ? or feel free to push as Chris is not available until much later

PVince81 commented:
seems CI died, there isn't even a job for this PR.
I'll trigger it with a rebase

PVince81 commented:
/rebase

blizzz committed:
fix querybuilder instance recuse

...and execute delete query

Signed-off-by: Arthur Schiwon <blizzz@arthur-schiwon.de>

nickvergessen commented on core/Migrations/Version25000Date20221007010957.php:
```diff
@@ -0,0 +1,73 @@
+		$updateQuery->update('preferences')
+			->set('appid', $updateQuery->createNamedParameter('theming'))
+			->where($updateQuery->expr()->eq('appid', $updateQuery->createNamedParameter('dashboard')))
+			->andWhere($orExpr);
```
Can not be reused as the variables are not set on the querybuilder object

blizzz replied:
sheesh, oversaw that one :grimacing:  thanks

nickvergessen committed:
Fix migration parameter handling

Signed-off-by: Joas Schilling <coding@schilljs.com>

PVince81 commented:
@blizzz did you retest after the latest changes ? just to confirm

blizzz commented:
> @blizzz did you retest after the latest changes ? just to confirm

No, had to rush to an appointment.

blizzz commented:
First off, I do not think it fixes the original issue, #34458, as in the description. On RC2 the settings are already stored with the "theming" appid.

```
sqlite> select * from oc_preferences where configkey in ('background', 'backgroundVersion');
admin|theming|background|custom
admin|theming|backgroundVersion|1
```

This fixes #34216 INSTEAD which is also great, no question. The concern of the other one are updates within the 25 series.

PVince81 commented:
@blizzz yes, we decided on purpose to "bother" people who are on RC3+ for the sake of simplifying/accelerating the migration (iterating over all users to find existing ones seemed like overkill). So only users post RC3 will have to re-set their backgrounds.

blizzz commented:
@PVince81 that's two different things. There will be an upgrade to 25.0.1 later on. Or is it ruled out, that (this patch not considered) it will not bug people when upgrading 25.0.0 → 25.0.1?

PVince81 commented:
@blizzz since this is a migration, it will only ever run once. So people upgrading from 25.0.0 to 25.0.x will already have it marked as run. I do agree that there's a slight concern in case someone would decided to clear the migrations or rerun this one.
Should we make two migrations out of it ?

There's also the option of ignoring the database conflict and only using "update", but then people post RC2->25.0.0 will get an update error because we can't overwrite already existing values, in case they exist. This is only for users who had a non-migrated dashboard wallpaper set and at the same time manually set a theme wallpaper.
Now wondering if it's even possible to get into this state ?

PVince81 commented:
for the sake of exercise, if we go for "update only" with a documented workaround, the query to cleanup existing dashboard would be as follow (untested):
```sql
delete from oc_preferences where appid="dashboard" where configkey in ('background', 'backgroundVersion') where exists (select 1 from oc_preferences where appid="theming" where configkey in ('background', 'backgroundVersion'))
```

blizzz commented:
Now i cannot reproduce https://github.com/nextcloud/server/issues/34458 anymore on a clean setup. I've had it on our instance.

PVince81 commented:
I've done a local test as follows:

1. Start with v24.0.6 with three users: alice, bob, charles
2. Login as alice and set a dashboard wallpaper
3. Login as bob and set a dashboard wallpaper
4. Update to v25.0.0 RC2
5. As alice, set a theming wallpaper

duplicates exist for bob:
```
MariaDB [nextcloud]> select * from oc_preferences where appid in ('theming', 'dashboard');
+--------+-----------+-------------------+-------------+
| userid | appid     | configkey         | configvalue |
+--------+-----------+-------------------+-------------+
| alice  | dashboard | background        | custom      |
| alice  | dashboard | backgroundVersion | 1           |
| alice  | dashboard | firstRun          | 0           |
| bob    | dashboard | background        | custom      |
| bob    | dashboard | backgroundVersion | 1           |
| bob    | dashboard | firstRun          | 0           |
| bob    | theming   | background        | custom      |
| bob    | theming   | backgroundVersion | 1           |
+--------+-----------+-------------------+-------------+
```

PVince81 commented:
@blizzz I've reproduced it, see last (edited) comment

blizzz commented:
> @blizzz I've reproduced it, see last (edited) comment

You reproduced https://github.com/nextcloud/server/issues/34216

PVince81 commented:
and because of those duplicates as in https://github.com/nextcloud/server/pull/34461#issuecomment-1277454538 we cannot simply run the update like in the first version of this PR:
```
MariaDB [nextcloud]> update oc_preferences set appid='dashboard' where configkey in ('background', 'backgroundVersion');
ERROR 1062 (23000): Duplicate entry 'bob-dashboard-background' for key 'PRIMARY'
```

so the proposal was to delete the backgrounds that were set post RC, here by Bob, by deleting the matching "theming" keys.

so that in the end only users (early adopters) will have to manually re-set the background

PVince81 commented:
@blizzz please check oc_preferences for my user on c.nc.com, my background is still black as I haven't re-selected one.
then we can see the DB state

blizzz commented:
> 

```
> select appid, configkey, configvalue from oc_preferences where configkey in ('background', 'backgroundVersion') and userid = '*******';
+---------+-------------------+-------------+
| appid   | configkey         | configvalue |
+---------+-------------------+-------------+
| theming | background        | custom      |
| theming | backgroundVersion | 5           |
+---------+-------------------+-------------+
```

blizzz commented:
> @blizzz please check oc_preferences for my user on c.nc.com, my background is still black as I haven't re-selected one. then we can see the DB state

I had selected on in rc2 which vanished in rc3 and was all black for me. Maybe pre-condition is coming from 24.

PVince81 commented:
I checked user "Alice" on my test system which would represent that test case and here the background is reset to default instead of black.

Maybe c.nc.com is missing that cloudy default pic ?

blizzz commented:
> I checked user "Alice" on my test system which would represent that test case and here the background is reset to default instead of black.
> 
> Maybe c.nc.com is missing that cloudy default pic ?

Nope.

PVince81 commented:
@blizzz or maybe one needs to upgrade through RC1, RC2, etc to reach that state.

also would need to check in which key the path to the custom bkg image is set, maybe that one can be broken as well ?

blizzz commented:
> @blizzz or maybe one needs to upgrade through RC1, RC2, etc to reach that state.
> 
> also would need to check in which key the path to the custom bkg image is set, maybe that one can be broken as well ?

Maybe RC 1 then at most. I went now from 24.0.6 (dark mode, non-default stock background) via rc2 (became default background, changed to custom picked from files) and the to RC3. It stayed. So if there was something with rc (or the beta), it does not really matter now. Will close https://github.com/nextcloud/server/issues/34458 and do another check with this PR to confirm  https://github.com/nextcloud/server/issues/34216 is fixed.

blizzz commented:
Now I took 24.0.6, uploaded a file and set this as background. Then upgrade to RC3 + this patch. Now I have a black background, so, does not really work.

blizzz commented:
Files need to be moved 

```
data/$APPDATA_DIR/dashboard/$USERID/background.jpg → data/$APPDATA_DIR/theming/$USERID/background.jpg
```

blizzz committed:
graceful background image handling

- fallback to background image from old location
- migrate background images to new location as insensitive job

Signed-off-by: Arthur Schiwon <blizzz@arthur-schiwon.de>

PVince81 commented:
adding the latest idea from internal discussion with @blizzz:
- add a fallback in BackgroundService that checks both locations in oc_preferences and also on disk
- have a time-insensitive (nightly) background job that migrates the oc_preferences keys and also the image file

there's some local WIP work towards this idea that will be pushed later

PVince81 commented:
update:

the fallback + background job for migration is only for the background file, not the config.
we still migrate the config in this PR

PVince81 commented:
I've redone this test:

1. Start with v24.0.6 with three users: alice, bob, charles
2. Login as alice and set a dashboard wallpaper from a local file
3. Login as bob and set a dashboard wallpaper from a local file
4. Upgrade to v25.0.0 RC2
5. As alice, set a theming wallpaper
6. As charles, set a theming wallpaper
7. Switch to this branch
8. Upgrade
9. Check all user's wallpaper

```
MariaDB [nextcloud]> select * from oc_preferences where appid in ('theming', 'dashboard');
+---------+-----------+-------------------+-------------+
| userid  | appid     | configkey         | configvalue |
+---------+-----------+-------------------+-------------+
| alice   | dashboard | firstRun          | 0           |
| alice   | theming   | background        | custom      |
| alice   | theming   | backgroundVersion | 2           |
| bob     | dashboard | firstRun          | 0           |
| bob     | theming   | background        | custom      |
| bob     | theming   | backgroundVersion | 1           |
| charles | dashboard | firstRun          | 0           |
+---------+-----------+-------------------+-------------+
```
in appdata:
```
dashboard
├── alice
│   └── background.jpg
└── bob
    └── background.jpg
theming
├── 1
│   ├── favIcon-core
│   ├── favIcon-dashboard
│   └── icon-core-filetypes_text.svg
├── alice
│   └── background.jpg
└── charles
    └── background.jpg

3 directories, 5 files
```

- Alice: wallpaper is visible :white_check_mark: 
- Bob: wallpaper is visible **through fallback**: :white_check_mark: 
- Charles: wallpaper is gone, this is by design as discussed for users who created in RC2 :white_check_mark: 

10. Run background jobs
11. Check tree again

```
dashboard
├── alice
└── bob
theming
├── 1
│   ├── favIcon-core
│   ├── favIcon-dashboard
│   └── icon-core-filetypes_text.svg
├── alice
│   └── background.jpg
├── bob
│   └── background.jpg
└── charles
    └── background.jpg

4 directories, 6 files
```

- Alice: wallpaper is visible :white_check_mark: 
- Bob: wallpaper is visible: :white_check_mark: 

Note: sadly I still need to shift+refresh to get the wallpapers to refresh, separate issue

**tldr; works as expected** :+1:

PVince81 commented:
- [ ]  raise a ticket to make the background job more scalable with a lot of users as we don't want to fetch a lot of users every time the job runs @blizzz
